### ISSUE TITLE
S3 Metadata name characters not preserved when run as server

### ISSUE THREAD

**Issue by thehesiod** [getmoto/moto#2054]
From what I see the server back-end sticks the metadata keys in an `EnvironHeaders` class, which then mucks with both the casing and certain characters like "_".

If I set a key called "src_etag", it gets returned "src-etag"  I believe the `tolower()` is correct, however the "_" -> "-" is not as it's breaking my unittests when moving to moto

**Comment by spulec** [comment]
Hmmm. Is this happening in moto itself or inside of flask/werkzeug?

We will want to find a fix either way, but it will be more annoying if the latter.

**Comment by thehesiod** [comment]
I believe it was in moto.

**Comment by spulec** [comment]
Hmmm, I'm not able to find any reference to `EnvironHeaders` in our codebase.

The following code is working for me. Can you confirm this is still a problem against `master`?

```python
@mock_s3_deprecated
@reduced_min_part_size
def test_multipart_upload_with_headers():
    conn = boto.connect_s3('the_key', 'the_secret')
    bucket = conn.create_bucket("foobar")

    multipart = bucket.initiate_multipart_upload(
        "the-key", metadata={"src_etag": "bar"})
    part1 = b'0' * 10
    multipart.upload_part_from_file(BytesIO(part1), 1)
    multipart.complete_upload()

    key = bucket.get_key("the-key")
    key.metadata.should.equal({"src_etag": "bar"})
```

**Comment by thehesiod** [comment]
@spulec note the title, you need to run moto as a server.  I just re-ran into this issue

**Comment by spulec** [comment]
(sorry for delay on this)

Ah, I see. Do you have code to reproduce?

**Comment by qqshfox** [comment]
Encountering similar issue here. Spawned the server inside of flask/werkzeug just like this https://github.com/ZwickyTransientFacility/alertbase/blob/87d642fbd2ba4d7c77640450f269820571192376/test/test_blobstore.py#L14

The root cause maybe similar to this https://github.com/localstack/localstack/issues/459#issuecomment-346756560.

**Comment by qqshfox** [comment]
Here's the code to reproduce:

```python
import boto3.session
import moto.server
import pytest
import threading
import werkzeug


TEST_BUCKET = 'test-bucket'


@pytest.fixture
def s3_server():
    ip = "127.0.0.1"
    port = "5287"
    app = moto.server.DomainDispatcherApplication(
        moto.server.create_backend_app,
        service="s3",
    )
    app.debug = True

    server = werkzeug.serving.make_server(ip, port, app, True)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    yield f"http://{ip}:{port}"
    server.shutdown()
    thread.join()


@pytest.fixture(scope='function')
def s3_session():
    return boto3.session.Session()


@pytest.fixture(scope='function')
def s3_client(s3_session, s3_server):
    return s3_session.client('s3', endpoint_url=s3_server)


@pytest.fixture(autouse=True)
def create_test_bucket(s3_client):
    s3_client.create_bucket(Bucket=TEST_BUCKET)

    yield


def test_metadata_with_underscore(s3_client):
    key = 'test-key'
    metadata = {
        'dash-key': 'dash-value',
        'underscore_key': 'underscore_value',
    }

    s3_client.put_object(Bucket=TEST_BUCKET, Key=key, Metadata=metadata)

    assert s3_client.head_object(Bucket=TEST_BUCKET, Key=key)["Metadata"] == metadata
```

**Comment by qqshfox** [comment]
It's the same as just using `moto_server` because `moto_server` uses `werkzeug` underlying as well.

**Comment by bblommers** [comment]
The `EnvironHeaders`-class comes from werkzeug: https://github.com/pallets/werkzeug/blob/master/src/werkzeug/datastructures.py#L1348

This issue has been raised before: https://github.com/pallets/werkzeug/issues/940

TLDR: This is how WSGI parses headers, by using dashes instead of underscores. I don't think Moto can do anything about this, besides completely rewriting MotoServer in a way that does not use werkzeug/WSGI

**Comment by thehesiod** [comment]
unfortunate, too bad there's not even hope of adding a param to change behavior it seems. If this work is to eventually be done I suggest swapping to something async, can help with running multiple services together in same context

**Comment by qqshfox** [comment]
looks like gunicorn has the same behavior:
https://github.com/benoitc/gunicorn/blob/master/gunicorn/http/wsgi.py#L136

**Comment by qqshfox** [comment]
waitress drops headers containing `_`:
https://github.com/Pylons/waitress/blob/master/src/waitress/parser.py#L231

```python
import waitress
import moto.server


ip = "127.0.0.1"
port = "5287"
app = moto.server.DomainDispatcherApplication(
    moto.server.create_backend_app,
    service="s3",
)
app.debug = True

waitress.serve(app, port=5287)
```

**Comment by qqshfox** [comment]
same as bjoern:
https://github.com/jonashaag/bjoern/blob/master/bjoern/request.c#L154

**Comment by qqshfox** [comment]
@bblommers @thehesiod it seems not many other alternatives out there. Do you guys have any suggestion on this?

**Comment by thehesiod** [comment]
I'd suggest aiohttp. It has a concept of raw headers

**Comment by thehesiod** [comment]
so latest version is pretty weird, it just totally strips all the metadata keys that have `_`.  Given amazon will always return `_` as `-` I'm going to close this.  annoying, but work-around is simple enough, swap to `-` to match what amazon gives you back

**Comment by thehesiod** [comment]
i figured out a workaround:
```python
def _werkzeug_make_environ(original_method, instance, args, kwargs):
    environ = original_method(*args, **kwargs)

    for key, value in instance.headers.items():
        key = key.upper().replace("-", "_")
        value = value.replace("\r\n", "")

        if key.startswith('X_AMZ_META_') and '_' in key:
            if key not in ("CONTENT_TYPE", "CONTENT_LENGTH"):
                key = f"HTTP_{key}"
                if key in environ:
                    value = f"{environ[key]},{value}"

            environ[key] = value

    return environ


@pytest.fixture(scope="session", autouse=True)
def default_function_fixture(request):
    for patch_path in [
        'datadog.dogstatsd.base.DogStatsd._send_to_server',
        'datadog.dogstatsd.base.statsd._send',
    ]:
        patcher = patch(patch_path)
        patcher.start()

        request.addfinalizer(patcher.stop)

    patcher = patch.dict(os.environ, {'UNITTEST': 'true'})
    patcher.start()
    request.addfinalizer(patcher.stop)

    request.addfinalizer(Client.close_shared_connectors)

    # TODO: this should happen via a __init__ file in the tests folder along with patching fbn_logging
    logging.basicConfig(level=logging.INFO)
    logging.captureWarnings(True)

    with wrapt_method(werkzeug.serving.WSGIRequestHandler, 'make_environ', _werkzeug_make_environ):
        yield
```
